### PR TITLE
Added optional `role` field to `VirtualDeviceContext`.

### LINKED ISSUE #6258: Add Role to VDC
As a Network Automation Engineer, I want to be able to specify a Role for a VDC on a device, so that I can capture and use automation context related to the role.

I will know this is complete when:
- VDCs assigned to a device can have an optional role

### PR DESCRIPTION
**Pull request by HanlinMiao**
# Closes #6258 
# What's Changed

Added optional `role` field to `VirtualDeviceContext`.

# Screenshots

# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/extras/tests/test_jobs.py
Comment (timizuoebideri1) [inline nautobot/extras/tests/test_jobs.py:1]: Maybe revert changes to this file?

Comment (HanlinMiao) [reply nautobot/extras/tests/test_jobs.py:1]: I think this is just a `ruff` fix

File: nautobot/dcim/tables/devices.py [hunk 1]
```diff
@@ -1411,3 +1412,48 @@ class Meta(BaseTable.Meta):
             "tags",
             "actions",
         )
+
+
+class VirtualDeviceContextTable(StatusTableMixin, BaseTable):
```
Comment (glennmatthews) [inline nautobot/dcim/tables/devices.py:1417]: Is this in conflict with #6209? Should this branch be based under that one?

Comment (HanlinMiao) [reply nautobot/dcim/tables/devices.py:1417]: I dont think this is in conflict, I pretty much copied the VirtualDeviceContextTable from #6029 and added the role column to it. But I do think #6029 should be merged first.

Comment (timizuoebideri1) [reply nautobot/dcim/tables/devices.py:1417]: I think its better your base branch is #6209

File: nautobot/dcim/models/devices.py [hunk 2]
```diff
@@ -1864,6 +1864,7 @@ class VirtualDeviceContext(PrimaryModel):
         null=True,
     )
     status = StatusField(blank=False, null=False)
+    role = RoleField(blank=True, null=True, related_name="virtual_device_contexts")
```
Comment (glennmatthews) [inline nautobot/dcim/models/devices.py:1867]: Shouldn't need an explicit `related_name` since RoleField uses `ForeignKeyWithAutoRelatedName`, I believe.

File: nautobot/cloud/factory.py [hunk 3]
```diff
@@ -69,7 +69,7 @@ class Meta:
     )
     cloud_account = random_instance(models.CloudAccount, allow_null=False)
     has_parent = NautobotBoolIterator()
-    extra_config = factory.Faker("pydict", value_types=[str, bool, int])
+    extra_config = factory.Faker("pydict", nb_elements=1, value_types=[str, bool, int])
```
Comment (glennmatthews) [inline nautobot/cloud/factory.py:72]: revert, doesn't seem to be fixing the CI failure.

File: nautobot/cloud/tests/test_api.py [hunk 4]
```diff
@@ -147,7 +147,7 @@ def setUpTestData(cls):
             "description": "A new description",
             "cloud_resource_type": cloud_resource_types[1].pk,
             "cloud_account": cloud_accounts[1].pk,
-            "extra_config": {"A": 1, "B": 2, "C": 3},
+            "extra_config": {"A": 1},
```
Comment (glennmatthews) [inline nautobot/cloud/tests/test_api.py:150]: revert, doesn't seem to be fixing the CI failure.

File: development/docker-compose.mysql.yml [hunk 5]
```diff
@@ -11,6 +11,7 @@ services:
       - "NAUTOBOT_DB_ENGINE=django_prometheus.db.backends.mysql"
   db:
     image: mysql:8
+    command: --sort-buffer-size=1M
```
Comment (glennmatthews) [inline development/docker-compose.mysql.yml:14]: Probably the wrong place to add this - this is for the local dev environment, not CI.

File: .github/workflows/ci_pullrequest.yml [hunk 6]
```diff
@@ -205,6 +205,7 @@ jobs:
     services:
       mysql:
         image: "mysql:8"
+        command: --sort-buffer-size=1M
```
Comment (glennmatthews) [inline .github/workflows/ci_pullrequest.yml:208]: Invalid GHA syntax, unfortunately. Not sure what the equivalent is.

File: nautobot/extras/migrations/0117_task_queue_to_job_queue_migration.py
Comment (glennmatthews) [inline nautobot/extras/migrations/0117_task_queue_to_job_queue_migration.py:1]: Why the changes to this file?

File: nautobot/dcim/tests/test_api.py [hunk 7]
```diff
@@ -3377,6 +3377,7 @@ class VirtualDeviceContextTestCase(APIViewTestCases.APIViewTestCase):
     def setUpTestData(cls):
         devices = Device.objects.all()
         vdc_status = Status.objects.get_for_model(VirtualDeviceContext)[0]
+        vdc_role = Role.objects.get_for_model(VirtualDeviceContext)[0]
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_api.py:3380]: Latest test failure *might* be due to not finding a valid Role object? Hard to say for sure.

File: .github/workflows/ci_pullrequest.yml [hunk 8]
```diff
@@ -227,7 +227,7 @@ jobs:
           poetry-version: "1.8.2"
           python-version: "3.12"
       - name: "Run unittest"
-        run: "poetry run invoke unittest --failfast --no-keepdb --no-cache-test-fixtures --parallel"
+        run: "poetry run invoke unittest --no-keepdb --no-cache-test-fixtures"
```
Comment (glennmatthews) [inline .github/workflows/ci_pullrequest.yml:230]: revert now that test is passing :)

File: nautobot/cloud/factory.py [hunk 9]
```diff
@@ -86,7 +86,7 @@ def prefixes(self, create, extracted, **kwargs):
             if extracted:
                 self.prefixes.set(extracted)
             else:
-                self.prefixes.set(get_random_instances(Prefix))
+                self.prefixes.set(get_random_instances(model_or_queryset_or_lambda=Prefix, maximum=1))
```
Comment (glennmatthews) [inline nautobot/cloud/factory.py:89]: can this be reverted?

File: nautobot/extras/migrations/0117_task_queue_to_job_queue_migration.py [hunk 10]
```diff
@@ -15,10 +14,7 @@ def migrate_task_queues_to_job_queues(apps, schema):
     # should always create default job queue
     JobQueue.objects.get_or_create(name="default", defaults={"queue_type": JobQueueTypeChoices.TYPE_CELERY})
     for job in Job.objects.all():
-        task_queues = job.task_queues or [settings.CELERY_TASK_DEFAULT_QUEUE]
-        # remove duplicates?
-        task_queues = list(set(task_queues))
-
+        task_queues = job.task_queues
```
Comment (glennmatthews) [inline nautobot/extras/migrations/0117_task_queue_to_job_queue_migration.py:17]: revert? Or is this a desirable change?

File: nautobot/ipam/tests/test_api.py [hunk 11]
```diff
@@ -1297,24 +1297,24 @@ def setUpTestData(cls):
         # make sure there are 4 locations without vlans 1 and 2 for the create_data below
         for i in range(4):
             cls.locations[i].vlans.set([])
-        locations_without_vlans = cls.locations.exclude(vlans__in=[cls.vlans[0], cls.vlans[1]])
+        locations_without_vlans = cls.locations.exclude(vlans__isnull=False)
```
Comment (glennmatthews) [inline nautobot/ipam/tests/test_api.py:1300]: Perhaps:

```suggestion
        locations_without_vlans = cls.locations.filter(vlans__isnull=True)
```

Comment (glennmatthews) [reply nautobot/ipam/tests/test_api.py:1300]: Are we guaranteed to have four locations with no associated VLANs? Or should we just explicitly create four such locations here to make the test more robust?

In either case, should the fixes in this file be backported to `develop` as well?

File: tasks.py [hunk 12]
```diff
@@ -754,7 +754,7 @@ def unittest(
     if verbose:
         command += " --verbosity 2"
     if parallel or label == "nautobot":
-        command += " --parallel"
+        # command += " --parallel"
```
Comment (glennmatthews) [inline tasks.py:757]: revert?

File: nautobot/ipam/tests/test_api.py (hunk 11 above)
Comment (HanlinMiao) [reply nautobot/ipam/tests/test_api.py:1300]: Yeah we were using index 1-4 which are expecting 5 locations it should be 0-3. Took me a while to figure out the obvious lol. I will backported it to develop

File: nautobot/extras/migrations/0117_task_queue_to_job_queue_migration.py (hunk 10 above)
Comment (HanlinMiao) [reply nautobot/extras/migrations/0117_task_queue_to_job_queue_migration.py:17]: This is a desirable change

File: nautobot/cloud/factory.py (hunk 9 above)
Comment (HanlinMiao) [reply nautobot/cloud/factory.py:89]: I dont think so. since this is the sort_buffer_size error fix.

Comment (HanlinMiao) [reply nautobot/cloud/factory.py:89]: Add todo comment

File: nautobot/ipam/tests/test_api.py (hunk 11 above)
Comment (glennmatthews) [reply nautobot/ipam/tests/test_api.py]: Ok - are we guaranteed to always have 4 suitable locations with the current implementation and factories?

Comment (HanlinMiao) [reply nautobot/ipam/tests/test_api.py]: Yes definitely, we will have more than 10 locations with VLAN content_type. And we are using a for loop above to make sure that we have 4.

**Review by glennmatthews — APPROVED** [review]

File: nautobot/cloud/factory.py [hunk 13]
```diff
@@ -86,6 +86,9 @@ def prefixes(self, create, extracted, **kwargs):
             if extracted:
                 self.prefixes.set(extracted)
             else:
+                # TODO Investigate https://github.com/nautobot/nautobot/actions/runs/11019738391/job/30603271529
```
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/cloud/factory.py:89]: https://github.com/nautobot/nautobot/issues/6301
